fat/fat2/addressbalancedeltamap.go

```diff
@@ -0,0 +1,78 @@
+package fat2
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/Factom-Asset-Tokens/factom"
+)
+
+type AddressBalanceDeltaMap map[factom.FAAddress]map[string]int64
```

Comment:
It is a representation of the resulting changes to balances if this transaction were to be executed.

So if address A converted 1 PEG to 5 pUSD, the balance delta map would look like:
```
deltas[A]["PEG"] == -1
deltas[A]["pUSD"] == 5
```

The intention was to allow the output calculation to be done within the fat2 transaction package, and not have to be reimplemented by anyone trying to consume the package.

Perhaps there's a better name for the type?

Reply:
Personally I am not convinced of the need for this. The database will compute the logic for the state transition and probably do it more efficiently than this. 

Have you read my latest addition to the fatd wiki on the relationship between state and the database? https://github.com/Factom-Asset-Tokens/fatd/wiki/2.-Database#state

I'm open to discussing this further but at the moment I would like to see the fat types only perform data validation outside of the context of state. I would like to see the state transition logic implemented in the database package.

Reply:
I disagree honestly. I think the transaction should contain all the logic necessary to evaluate a conversion. If I am constructing a conversion transaction in one application (outside of fatd), I shouldn't have to import another application's database package if I want to evaluate what that transaction's outputs are given a set of conversion rates.

For fat0/1 transactions it makes sense because the transaction itself is already a list of deltas (what to add/subtract from addresses explicitly). However in pegnet, where you don't know the conversion rates at the time of submission, the results of a transaction must be computed. Not including it in the fat2 package opens us up to more opportunities for anyone that needs to manually compute what a conversion results in.

Reply:
Additionally, this function `GetBalanceDeltas()` does serve as data validation, as it computes whether there are enough inputs to cover the requested of outputs. Similar to how you check for `t.Inputs.Sum() != t.Outputs.Sum()` inside fat0's `Transaction.ValidData()` function.

However, in addition to performing the data validation, it also will return what the result of the transaction is because it had to compute the results anyways

Reply:
>  I think the transaction should contain all the logic necessary to evaluate a conversion.

I also wanted to achieve this initially, but found it resulted in weird inherent dependencies between the database connection and the fat packages. Or it required an abstraction layer that would require loading any necessary state into memory first, which reduces our ability to leverage the database for certain things (schema constraints for example). 

It's worth asking who is the intended consumer of this package? In the case of the existing fat packages, they are intended to be imported by the engine, database, srv, and any clients, like the fat-cli packages. The only package that is truly aware of state is the database. Any client for the fat packages will not want to rebuild state. They will want to ask fatd for it. When the srv package needs to evaluate a transaction, it needs to ask the database (which it gets a handle to from the engine package). You check the balances you are interested to ensure that you have enough for the inputs while constructing a transaction. That's all they need.

But even if you did want your client to rebuild state, then you would not want to inherit fatd's database design choices anyway. You'd want to use the same data validation, but you'd have your own ideas about how to implement the database. 

Consider that you want the state transition/validation rules to be some of the easiest code to understand and you want it to only exist in one place really. 

At least for now, lets write these packages with the engine and database and srv in mind, before considering external clients. 

> However in pegnet, where you don't know the conversion rates at the time of submission, input/outputs' values are often implicit and the results of a transaction must be computed. 

This kind of computation strikes me as part of the state transition. If the current state is needed to evaluate a conversion then you will need to somehow bridge that gap to the database.

> Additionally, this function GetBalanceDeltas() does serve as data validation, as it computes whether there are enough inputs to cover the requested of outputs.

Okay, and that is definitely important to check. You want the database to be able to assume that the structure of this object is completely valid after calling Validate() and getting nil error. Maybe this is an important helper function for the database.

But before we make any final decisions I need to understand these transactions better from the pegnet side.

Reply:
No longer has a `GetBalanceDeltas()` since conversions were stripped of many features that were previously requested. No need for the helper function anymore